### Running in Docker container on EC2 instance

#### Started Docker container with the following command:

```docker run  -p 8888:8888 -v /home/ubuntu/gitrepos/LabDocs/jupyter_nbs/sam:/home/notebooks -v /home/ubuntu/data:/data/ -it kubu4/bioinformatics:v11 /bin/bash```

The command allows access to Jupyter Notebook over port 8888 and makes my Jupyter Notebook GitHub repo and my data files accessible to the Docker container

Once the container was started, started Jupyter Notebook with the following command inside the Docker container:

```jupyter notebook```

This is configured in the Docker container to launch a Jupyter Notebook without a browser on port 8888.

#### Created a tunnel from my local computer to the Docker container:

```ssh -i ~/Dropbox/Lab/Sam/bioinformatics.pem -N -L localhost:8888:localhost:8888 ubuntu@ec2.ip.address```

This command is run in a separate Terminal window than the one that is used to ssh into the EC2 instance to start Docker and all of that.

This ssh command specifies to use my Amazon EC2 authentication file (bioinformatics.pem), along with the -N and -L options for port forwarding stuff (see man ssh for deets), and binds the port 8888 on my local computer to port 8888 on the EC2 isntance. 

The tunnel allows me to start the Jupyter Notebook in my web browser. I enter ```localhost:8888``` in as the URL.

In [1]:
%%bash
date

Fri Jul 15 16:56:46 UTC 2016


In [2]:
%%bash
hostname

570c28713283


### Check computer specs

In [3]:
%%bash
lscpu

Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 63
Model name:            Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz
Stepping:              2
CPU MHz:               2900.088
BogoMIPS:              5800.17
Hypervisor vendor:     Xen
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              25600K
NUMA node0 CPU(s):     0-7


In [4]:
cd /data/

/data


In [5]:
ls

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_1.fq.gz  1NF_25A_2.fq.gz*
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_2.fq.gz  1NF_26A_1.fq.gz*
1HL_10A_1.fq.gz*                                            1NF_26A_2.fq.gz*
1HL_10A_2.fq.gz*                                            1NF_27A_1.fq.gz*
1HL_11A_1.fq.gz*                                            1NF_27A_2.fq.gz*
1HL_11A_2.fq.gz*                                            1NF_28A_1.fq.gz*
1HL_12A_1.fq.gz*                                            1NF_28A_2.fq.gz*
1HL_12A_2.fq.gz*                                            1NF_29A_1.fq.gz*
1HL_13A_1.fq.gz*                                            1NF_29A_2.fq.gz*
1HL_13A_2.fq.gz*                                            1NF_2A_1.fq.gz*
1HL_14A_1.fq.gz*                                            1NF_2A_2.fq.gz*
1HL_14A_2.fq.gz*                                            1NF_30A_1.fq.gz*
1HL_15A_1.fq.gz*                                            1NF_30A_2.fq.gz*
1

### Rename FASTQ files to match R1 and R2 requirements for pyrad demultiplexing¶

In [9]:
%%bash
mv 160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_1.fq.gz 160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz

In [10]:
%%bash
mv 160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_2.fq.gz 160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz

In [11]:
cd analysis/

/data/analysis


In [12]:
ls

barcodes.txt  params.txt


In [13]:
%%bash
cat params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
./                     ## 1. Working directory                                 (all)
/data/oly_gbs/*.gz             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/analysis/20160609_pyrad/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
16                     ## 7. N processors (parallel)                           (all)
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see l

In [14]:
%%bash
mv /data/160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz /data/analysis/

In [16]:
%%bash
mv /data/160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz /data/analysis/

In [17]:
ls

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz  barcodes.txt
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz  params.txt


Need to edit params file to match current locations of data files. Will do outside of notebook. BRB...

In [18]:
%%bash
cat params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
16                     ## 7. N processors (parallel)                           (all)
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see l

### View barcodes.txt file

In [19]:
%%bash
cat barcodes.txt

1NF_1A	CTCC
1NF_2A	TGCA
1NF_4A	ACTA
1NF_5A	CAGA
1NF_6A	AACT
1NF_7A	GCGT
1NF_8A	CGAT
1NF_9A	GTAA
1NF_10A	AGGC
1NF_11A	GATC
1NF_12A	TCAC
1NF_13A	TGCGA
1NF_14A	CGCTT
1NF_15A	TCACC
1NF_16A	CTAGC
1NF_17A	ACAAA
1NF_18A	TTCTC
1NF_19A	AGCCC
1NF_20A	GTATT
1NF_21A	CTGTA
1NF_22A	ACCGT
1NF_23A	GCTTA
1NF_24A	GGTGT
1NF_25A	AGGAT
1NF_26A	ATTGA
1NF_27A	CATCT
1NF_28A	CCTAC
1NF_29A	GAGGA
1NF_30A	GGAAC
1NF_31A	GTCAA
1NF_32A	TAATA
1NF_33A	TACAT
1SN_1A	TCGTT
1SN_2A	GGTTGT
1SN_3A	CCAGCT
1SN_4A	TTCAGA
1SN_5A	TAGGAA
1SN_6A	GCTCTA
1SN_7A	CCACAA
1SN_8A	CTTCCA
1SN_9A	GAGATA
1SN_10A	ATGCCT
1SN_11A	AGTGGA
1SN_12A	ACCTAA
1SN_13A	ATATGT
1SN_14A	ATCGTA
1SN_15A	CATCGT
1SN_16A	CGCGGT
1SN_17A	CTATTA
1SN_18A	GCCAGT
1SN_19A	GGAAGA
1SN_20A	GTACTT
1SN_21A	GTTGAA
1SN_22A	TAACGA
1SN_23A	TGGCTA
1SN_24A	TATTTTT
1SN_25A	CTTGCTT
1SN_26A	ATGAAAC
1SN_27A	AAAAGTT
1SN_28A	GAATTCA
1SN_29A	GAACTTC
1SN_30A	GGACCTA
1SN_31A	GTCGATT
1SN_32A	AACGCCT
1HL_1A	AATATGC
1HL_2A	ACGTGTT
1HL_3A	ATTAATT
1HL_4A	ATTGGAT
1HL_5A	CATAAGT
1HL_6A	CGCTGAT
1H

### Step 1: De-multiplex reads

In [20]:
%%bash
time pyrad -p params.txt -s 1


	checking the file to make sure it properly demultiplexes

no stats file generated




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 1: sorting reads by barcode
	 Traceback (most recent call last):
  File "/usr/local/bin/pyrad", line 9, in <module>
    load_entry_point('pyrad==3.0.66', 'console_scripts', 'pyrad')()
  File "build/bdist.linux-x86_64/egg/pyrad/pyRAD.py", line 292, in main
  File "build/bdist.linux-x86_64/egg/pyrad/sortandcheck2.py", line 428, in main
  File "build/bdist.linux-x86_64/egg/pyrad/sortandcheck2.py", line 389, in writefunc
ValueError: max() arg is an empty sequence

real	93m3.141s
user	81m1.775s
sys	0m19.037s


Hmmm, will re-run this since it seems to have failed...

In [21]:
%%bash
time pyrad -p params.txt -s 1


	fastq/ directory in working directory contains data, move/remove it before running step 1





     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


real	0m2.740s
user	0m0.304s
sys	0m0.115s


In [23]:
cd fastq/

/data/analysis/fastq


In [25]:
ls

In [26]:
cd ..

/data/analysis


In [27]:
ls

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz  fastq/
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz  params.txt
barcodes.txt                                                  stats/


In [28]:
%%bash
rm -rf fastq/

In [29]:
ls

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz  params.txt
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz  stats/
barcodes.txt


In [ ]:
%%bash
time pyrad -p params.txt -s 1

Jupyter kernel crased while this was running, so no output to screen. I used the ```top``` command to verify that pyrad was still running despite Jupyter being down. Let's look at some files to see what happened...

In [1]:
pwd

u'/home/notebooks'

In [2]:
cd /data/analysis/

/data/analysis


In [3]:
%%bash
head -n 100 stats/s1.sorting.txt

file    	Nreads	cut_found	bar_matched


sample	true_bar	obs_bars	N_obs



In [4]:
ls

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz  fastq/
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz  params.txt
barcodes.txt                                                  stats/


In [5]:
ls fastq/

Guess it didn't complete...

In [6]:
rmdir fastq/

rmdir: failed to remove 'fastq/': Directory not empty


In [7]:
cd fastq/

/data/analysis/fastq


In [8]:
ls

In [9]:
rm -rf fastq/

In [10]:
ls

In [11]:
pwd

u'/data/analysis/fastq'

In [12]:
cd ..

/data/analysis


In [13]:
rm -rf fastq/

In [14]:
ls

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz  params.txt
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz  stats/
barcodes.txt


In [15]:
cd stats/

/data/analysis/stats


In [16]:
ls

s1.sorting.txt


In [17]:
rm s1.sorting.txt

In [18]:
cd ..

/data/analysis


In [19]:
%%bash
time pyrad -p params.txt -s 1


	checking the file to make sure it properly demultiplexes




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 1: sorting reads by barcode
	 ./bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: write error: No space left on device
/bin/cat: w

Let's see if the de-multiplexing really failed or not. Based on the run time, it seems like things should be fine.

In [1]:
cd /data/analysis/fastq/

/data/analysis/fastq


In [2]:
ls

1HL_10A_R1.fq.gz  1HL_35A_R1.fq.gz  1NF_25A_R1.fq.gz  1SN_18A_R1.fq.gz
1HL_10A_R2.fq.gz  1HL_35A_R2.fq.gz  1NF_25A_R2.fq.gz  1SN_18A_R2.fq.gz
1HL_11A_R1.fq.gz  1HL_3A_R1.fq.gz   1NF_26A_R1.fq.gz  1SN_19A_R1.fq.gz
1HL_11A_R2.fq.gz  1HL_3A_R2.fq.gz   1NF_26A_R2.fq.gz  1SN_19A_R2.fq.gz
1HL_12A_R1.fq.gz  1HL_4A_R1.fq.gz   1NF_27A_R1.fq.gz  1SN_1A_R1.fq.gz
1HL_12A_R2.fq.gz  1HL_4A_R2.fq.gz   1NF_27A_R2.fq.gz  1SN_1A_R2.fq.gz
1HL_13A_R1.fq.gz  1HL_5A_R1.fq.gz   1NF_28A_R1.fq.gz  1SN_20A_R1.fq.gz
1HL_13A_R2.fq.gz  1HL_5A_R2.fq.gz   1NF_28A_R2.fq.gz  1SN_20A_R2.fq.gz
1HL_14A_R1.fq.gz  1HL_6A_R1.fq.gz   1NF_29A_R1.fq.gz  1SN_21A_R1.fq.gz
1HL_14A_R2.fq.gz  1HL_6A_R2.fq.gz   1NF_29A_R2.fq.gz  1SN_21A_R2.fq.gz
1HL_15A_R1.fq.gz  1HL_7A_R1.fq.gz   1NF_2A_R1.fq.gz   1SN_22A_R1.fq.gz
1HL_15A_R2.fq.gz  1HL_7A_R2.fq.gz   1NF_2A_R2.fq.gz   1SN_22A_R2.fq.gz
1HL_16A_R1.fq.gz  1HL_8A_R1.fq.gz   1NF_30A_R1.fq.gz  1SN_23A_R1.fq.gz
1HL_16A_R2.fq.gz  1HL_8A_R2.fq.gz   1NF_30A_R2.fq.gz  1SN_23A_R2.fq.gz
1HL_17A_

In [3]:
ls -1 | wc -l

192


Correct number of files were produced from demultiplexing, so things look good on that front. Let's just glance at the stats file to see what it looks like...

NOTE: Bash trick above. The ls command above is using the number one (1), NOT lowercase L (l). This provides an accurate line count when you pipe the output to the word count command (wc).

In [4]:
%%bash
head ../stats/s1.sorting.txt

file    	Nreads	cut_found	bar_matched
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_.fq.gz	364261046	355811046	354725421


sample	true_bar	obs_bars	N_obs
1SN_27A    	AAAAGTT    	AAAAGTT	3325108   
1SN_27A    	AAAAGTT    	NAAAGTT	8330      
1SN_27A    	AAAAGTT    	AAAAATT	4748      
1SN_27A    	AAAAGTT    	ANAAGTT	1455      
1SN_27A    	AAAAGTT    	AAAATTT	782       


Things look OK to me. Next step!

### Merge paired-reads

The first thing we have to do is unzip the reads if they are gzipped. If you demultiplexed your data in pyrad they will be in the fastq/ directory.

In [5]:
%%bash
time gunzip /data/analysis/fastq/*.fq.gz


gzip: /data/analysis/fastq/1HL_19A_R2.fq.gz: unexpected end of file

real	52m44.818s
user	1m24.977s
sys	0m13.352s


In [7]:
%%bash
time for gfile in /data/analysis/fastq/*_R1.fq;
    do /usr/local/bioinformatics/pear-0.9.6-bin-64/pear-0.9.6-bin-64 -f $gfile \
            -r ${gfile/_R1.fq/_R2.fq} \
            -o ${gfile/_R1.fq/} \
            -n 33 \
            -t 33 \
            -q 10 \
            -j 20  >> pear.log 2>&1;
done


real	25m34.538s
user	71m32.940s
sys	0m17.129s


#### Set the data location to the de-multiplexed & merged data

In [8]:
%%bash
## set location of demultiplexed data that are 'pear' filtered
sed -i '/## 18./c\/data/analysis/fastq/*.assembled.fastq ## 18. demulti data loc ' /data/analysis/params.txt

In [9]:
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
16                     ## 7. N processors (parallel)                           (all)
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see l

#### Set the datatype to "merged"

In [10]:
%%bash
sed -i '/## 11./c\merged            ## 11. data type ' /data/analysis/params.txt

In [11]:
%%bash
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
16                     ## 7. N processors (parallel)                           (all)
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see l

### Step 2: Quality filtering

In [14]:
%%bash
time pyrad -p /data/analysis/params.txt -s 2



     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------

	sorted .fastq from /data/analysis/fastq/*.assembled.fastq being used
	step 2: editing raw reads 
	........
real	14m52.577s
user	39m28.511s
sys	0m7.845s


#### Let's take a look at the results

In [1]:
%%bash
cat /data/analysis/stats/s2.rawedit.txt

sample 	Nreads	passed	passed.w.trim	passed.total
1HL_10A.assembled	2890999	2127357	214294	2341651
1HL_11A.assembled	3088425	2242089	246561	2488650
1HL_12A.assembled	2323231	1728934	163227	1892161
1HL_13A.assembled	2701305	2075000	175929	2250929
1HL_14A.assembled	1865526	1407146	129089	1536235
1HL_15A.assembled	1837684	1342114	139898	1482012
1HL_16A.assembled	2100838	1568118	147901	1716019
1HL_17A.assembled	3106139	2259592	237499	2497091

    Nreads = total number of reads for a sample
    passed = retained reads that passed quality filtering at full length
    passed.w.trim= retained reads that were trimmed due to detection of adapters
    passed.total  = total kept reads of sufficient length
    note: you can set the option in params file to include trimmed reads of xx length. 


### Step 3: Clustering

In [2]:
%%bash
time pyrad -p /data/analysis/params.txt -s 3



     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	de-replicating files for clustering...

	step 3: within-sample clustering of 8 samples at 
	        '.88' similarity. Running 8 parallel jobs
	 	with up to 16 threads per job. If needed, 
		adjust to avoid CPU and MEM limits

Process Worker-3:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "build/bdist.linux-x86_64/egg/pyrad/potpour.py", line 31, in run
    res = self.func(*job)
  File "build/bdist.linux-x86_64/egg/pyrad/cluster7dp.py", line 545, in final
    alignwrap(outfolder+"/"+handle.split("/")[-1].replace(".edit",".clust.gz"), mindepth, muscle, w1)
  File "build/bdist.linux-x86_64/egg/pyrad/cluster7dp.py", line 415, in alignwrap
    stringnames = alignfast(names[0:200],seqs[0:200],muscle)
  Fi

Possibly overloading the CPUs. Let's change to reflect the actual number of cores (8).

In [4]:
%%bash
sed -i '/## 7. /c\8                   ## 7. N processors   ' /data/analysis/params.txt

In [5]:
%%bash
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
8                   ## 7. N processors   
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see line 20)(s2)
.88                    ## 10. W

Let's try this again...

In [6]:
%%bash
time pyrad -p /data/analysis/params.txt -s 3

	skipping 1HL_17A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_13A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_16A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_14A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_15A.assembled.clustS.gz already exists in /data/analysis/clust.88




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	de-replicating files for clustering...

	step 3: within-sample clustering of 8 samples at 
	        '.88' similarity. Running 8 parallel jobs
	 	with up to 16 threads per job. If needed, 
		adjust to avoid CPU and MEM limits

Process Worker-6:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "build/bdist.linux-x86_64/egg/pyrad/potpour.py", line 31, in run
    res = self.func(*job)
  File "build/bdist.linux-x86_64/egg/pyrad/cluster7dp.py", line 545, in final
    alignwrap(outfolder+"/"+handle.split("/")[-1].replace(".edit",".clust.gz"), mindepth, muscle, w1)
  File "build/bdist.linux-x86_64/egg/pyrad/cluster7dp.py", line 415, in alignwrap
    stringnames = alignfast(names[0:200],seqs[0:200],muscle)
  Fi

Well, I think the problem is that I'm running this at the same time as I'm running ustacks (a different program). It seems that ustacks is using all of the EC2 instance's resources (viewed using the Bash program "htop"). I will wait until that has completed and then run PyRad (or, upgrade the EC2 instance to a a greater number of CPUs or clone this instance and run two separate instances).

![](http://eagle.fish.washington.edu/Arabidopsis/20160718_ec2_ustacks_cpus.png)

In [1]:
%%bash
time pyrad -p /data/analysis/params.txt -s 3

	skipping 1HL_17A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_10A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_13A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_12A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_16A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_14A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_15A.assembled.clustS.gz already exists in /data/analysis/clust.88




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	de-replicating files for clustering...

	step 3: within-sample clustering of 8 samples at 
	        '.88' similarity. Running 8 parallel jobs
	 	with up to 16 threads per job. If needed, 
		adjust to avoid CPU and MEM limits

	sample 1HL_11A finished, 125407 loci

real	63m55.147s
user	44m17.743s
sys	29m14.039s


#### Let's take a look at the results

In [4]:
%%bash
cat /data/analysis/stats/s3.clusters.txt


taxa	total	dpt.me	dpt.sd	d>5.tot	d>5.me	d>5.sd	badpairs

    ## total = total number of clusters, including singletons
    ## dpt.me = mean depth of clusters
    ## dpt.sd = standard deviation of cluster depth
    ## >N.tot = number of clusters with depth greater than N
    ## >N.me = mean depth of clusters with depth greater than N
    ## >N.sd = standard deviation of cluster depth for clusters with depth greater than N
    ## badpairs = mismatched 1st & 2nd reads (only for paired ddRAD data)

HISTOGRAMS

    

taxa	total	dpt.me	dpt.sd	d>5.tot	d>5.me	d>5.sd	badpairs

    ## total = total number of clusters, including singletons
    ## dpt.me = mean depth of clusters
    ## dpt.sd = standard deviation of cluster depth
    ## >N.tot = number of clusters with depth greater than N
    ## >N.me = mean depth of clusters with depth greater than N
    ## >N.sd = standard deviation of cluster depth for clusters with depth greater than N
    ## badpairs = mismatched 1st & 2nd reads (only for p

### Step 4: Parameter estimation

In [6]:
%%bash
time pyrad -p /data/analysis/params.txt -s 4



     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 4: estimating error rate and heterozygosity
	........
real	34m34.339s
user	60m29.346s
sys	0m0.623s


In [7]:
%%bash
cat /data/analysis/stats/Pi_E_estimate.txt

taxa	H	E
1HL_13A.assembled	0.01234178	0.00144187	
1HL_10A.assembled	0.01399484	0.00139753	
1HL_15A.assembled	0.01242723	0.0011879	
1HL_12A.assembled	0.01240698	0.00157456	
1HL_16A.assembled	0.01381188	0.00054237	
1HL_17A.assembled	0.01332716	0.00127763	
1HL_14A.assembled	0.01266762	0.00203182	
1HL_11A.assembled	0.01216978	0.00165352	


Yeesh, this isn't right. Not all of the individuals are here. Should've noticed this before! Going back to unzipping files before merging paired reads.

In [8]:
%%bash
time gunzip /data/analysis/fastq/*.fq.gz


gzip: /data/analysis/fastq/1HL_19A_R2.fq.gz: unexpected end of file

real	0m35.063s
user	0m4.390s
sys	0m0.746s


In [9]:
ls -lh /data/analysis/fastq/

total 49G
-rw-r--r-- 1 root root 762M Jul 18 16:17 1HL_10A.assembled.fastq
-rw-r--r-- 1 root root 384M Jul 18 16:17 1HL_10A.discarded.fastq
-rw-r--r-- 1 root root    0 Jul 18 16:14 1HL_10A.unassembled.forward.fastq
-rw-r--r-- 1 root root    0 Jul 18 16:14 1HL_10A.unassembled.reverse.fastq
-rw-r--r-- 1 root root 969M Jul 17 06:24 1HL_10A_R1.fq
-rw-r--r-- 1 root root 1.0G Jul 17 06:25 1HL_10A_R2.fq
-rw-r--r-- 1 root root 822M Jul 18 16:22 1HL_11A.assembled.fastq
-rw-r--r-- 1 root root 433M Jul 18 16:22 1HL_11A.discarded.fastq
-rw-r--r-- 1 root root    0 Jul 18 16:17 1HL_11A.unassembled.forward.fastq
-rw-r--r-- 1 root root    0 Jul 18 16:17 1HL_11A.unassembled.reverse.fastq
-rw-r--r-- 1 root root 1.1G Jul 17 06:25 1HL_11A_R1.fq
-rw-r--r-- 1 root root 1.1G Jul 17 06:26 1HL_11A_R2.fq
-rw-r--r-- 1 root root 612M Jul 18 16:26 1HL_12A.assembled.fastq
-rw-r--r-- 1 root root 299M Jul 18 16:26 1HL_12A.discarded.fastq
-rw-r--r-- 1 root root    0 Jul 18 16:22 1HL_12A.unassembled.forward.fastq
-rw-r

Well, it turns out the the original demultiplexing actually failed. Many fq.gz files that are empty (i.e. 0 megabytes in size). Need to re-run demultiplexing.

#### Need to change data type back to pairgbs.

In [10]:
%%bash
sed -i '/## 11./c\pairgbs            ## 11. data type ' /data/analysis/params.txt

In [12]:
%%bash 
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
8                   ## 7. N processors   
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see line 20)(s2)
.88                    ## 10. W

In [13]:
%%bash
ls /data/analysis/

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz
barcodes.txt
clust.88
edits
fastq
params.txt
stats


In [14]:
%%bash
rm -rf /data/analysis/fastq/

In [15]:
%%bash
ls /data/analysis/

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz
barcodes.txt
clust.88
edits
params.txt
stats


In [16]:
cd /data/analysis/

/data/analysis


De-multiplex reads (again)

In [17]:
%%bash
time pyrad -p /data/analysis/params.txt -s 1

	skipping step 1: line 18 of input file shows seqs already sorted




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


real	0m0.225s
user	0m0.202s
sys	0m0.023s


In [18]:
%%bash
## set location of demultiplexed data that are 'pear' filtered
sed -i '/## 18./c\ ## 18. demulti data loc ' /data/analysis/params.txt

In [19]:
%%bash 
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
8                   ## 7. N processors   
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see line 20)(s2)
.88                    ## 10. W

In [20]:
%%bash
time pyrad -p /data/analysis/params.txt -s 1


	checking the file to make sure it properly demultiplexes




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 1: sorting reads by barcode
	 .
real	771m53.005s
user	687m46.614s
sys	3m12.997s


Let's check the output and make sure the files actually contain data...

In [22]:
ls -lh /data/analysis/fastq/

total 35G
-rw-r--r-- 1 root root 183M Jul 20 08:30 1HL_10A_R1.fq.gz
-rw-r--r-- 1 root root 226M Jul 20 08:31 1HL_10A_R2.fq.gz
-rw-r--r-- 1 root root 200M Jul 20 08:31 1HL_11A_R1.fq.gz
-rw-r--r-- 1 root root 247M Jul 20 08:32 1HL_11A_R2.fq.gz
-rw-r--r-- 1 root root 146M Jul 20 08:32 1HL_12A_R1.fq.gz
-rw-r--r-- 1 root root 181M Jul 20 08:32 1HL_12A_R2.fq.gz
-rw-r--r-- 1 root root 159M Jul 20 08:33 1HL_13A_R1.fq.gz
-rw-r--r-- 1 root root 197M Jul 20 08:33 1HL_13A_R2.fq.gz
-rw-r--r-- 1 root root 117M Jul 20 08:33 1HL_14A_R1.fq.gz
-rw-r--r-- 1 root root 146M Jul 20 08:34 1HL_14A_R2.fq.gz
-rw-r--r-- 1 root root 123M Jul 20 08:34 1HL_15A_R1.fq.gz
-rw-r--r-- 1 root root 154M Jul 20 08:34 1HL_15A_R2.fq.gz
-rw-r--r-- 1 root root 133M Jul 20 08:35 1HL_16A_R1.fq.gz
-rw-r--r-- 1 root root 166M Jul 20 08:35 1HL_16A_R2.fq.gz
-rw-r--r-- 1 root root 201M Jul 20 08:36 1HL_17A_R1.fq.gz
-rw-r--r-- 1 root root 252M Jul 20 08:36 1HL_17A_R2.fq.gz
-rw-r--r-- 1 root root 202M Jul 20 08:37 1HL_19A_R1.fq.gz
-rw-

And, let's make sure there's the expected number of individuals (96 individuals, sequenced from each end, so 192 total files)...

In [23]:
ls -1 /data/analysis/fastq/ | wc -l

192


#### Merge paired reads

The first thing we have to do is unzip the reads if they are gzipped. If you demultiplexed your data in pyrad they will be in the fastq/ directory.

In [24]:
%%bash
time gunzip /data/analysis/fastq/*.fq.gz


real	145m40.754s
user	14m53.245s
sys	2m6.261s


In [25]:
ls /data/analysis/stats/

Pi_E_estimate.txt  s1.sorting.txt  s2.rawedit.txt  s3.clusters.txt


In [26]:
ls /data/analysis/

160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R1_.fq.gz  edits/
160123_I132_FCH3YHMBBXX_L4_OYSzenG1AAD96FAAPEI-109_R2_.fq.gz  fastq/
barcodes.txt                                                  params.txt
clust.88/                                                     stats/


In [27]:
%%bash
time for gfile in /data/analysis/fastq/*_R1.fq;
    do /usr/local/bioinformatics/pear-0.9.6-bin-64/pear-0.9.6-bin-64 -f $gfile \
            -r ${gfile/_R1.fq/_R2.fq} \
            -o ${gfile/_R1.fq/} \
            -n 33 \
            -t 33 \
            -q 10 \
            -j 20  >> pear.log 2>&1;
done


real	173m24.008s
user	940m23.125s
sys	2m39.123s


In [29]:
ls -lh /data/analysis/fastq/

total 263G
-rw-r--r-- 1 root root  762M Jul 20 16:53 1HL_10A.assembled.fastq
-rw-r--r-- 1 root root  384M Jul 20 16:53 1HL_10A.discarded.fastq
-rw-r--r-- 1 root root     0 Jul 20 16:52 1HL_10A.unassembled.forward.fastq
-rw-r--r-- 1 root root     0 Jul 20 16:52 1HL_10A.unassembled.reverse.fastq
-rw-r--r-- 1 root root  969M Jul 20 08:30 1HL_10A_R1.fq
-rw-r--r-- 1 root root  1.0G Jul 20 08:31 1HL_10A_R2.fq
-rw-r--r-- 1 root root  822M Jul 20 16:55 1HL_11A.assembled.fastq
-rw-r--r-- 1 root root  433M Jul 20 16:55 1HL_11A.discarded.fastq
-rw-r--r-- 1 root root     0 Jul 20 16:53 1HL_11A.unassembled.forward.fastq
-rw-r--r-- 1 root root     0 Jul 20 16:53 1HL_11A.unassembled.reverse.fastq
-rw-r--r-- 1 root root  1.1G Jul 20 08:31 1HL_11A_R1.fq
-rw-r--r-- 1 root root  1.1G Jul 20 08:32 1HL_11A_R2.fq
-rw-r--r-- 1 root root  612M Jul 20 16:57 1HL_12A.assembled.fastq
-rw-r--r-- 1 root root  299M Jul 20 16:57 1HL_12A.discarded.fastq
-rw-r--r-- 1 root root     0 Jul 20 16:55 1HL_12A.unassembled.for

#### Set the data location to the de-multiplexed & merged data

In [30]:
%%bash
## set location of demultiplexed data that are 'pear' filtered
sed -i '/## 18./c\/data/analysis/fastq/*.assembled.fastq ## 18. demulti data loc ' /data/analysis/params.txt

In [31]:
%%bash
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
8                   ## 7. N processors   
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see line 20)(s2)
.88                    ## 10. W

#### Set the datatype to "merged"

In [32]:
%%bash
sed -i '/## 11./c\merged            ## 11. data type ' /data/analysis/params.txt

In [33]:
%%bash
cat /data/analysis/params.txt

==** parameter inputs for pyRAD version 3.0.66  **======================== affected step ==
/data/analysis/                     ## 1. Working directory                                 (all)
/data/analysis/             	       ## 2. Loc. of non-demultiplexed files (if not line 18)  (s1)
/data/analysis/barcodes.txt	               ## 3. Loc. of barcode file (if not line 18)             (s1)
/usr/local/bioinformatics/vsearch-1.11.1-linux-x86_64/bin/vsearch                ## 4. command (or path) to call vsearch (or usearch)    (s3,s6)
/usr/local/bioinformatics/muscle3.8.31_i86linux64                    ## 5. command (or path) to call muscle                  (s3,s7)
CWGC                   ## 6. Restriction overhang (e.g., C|TGCAG -> TGCAG)     (s1,s2)
8                   ## 7. N processors   
6                      ## 8. Mindepth: min coverage for a cluster              (s4,s5)
4                      ## 9. NQual: max # sites with qual < 20 (or see line 20)(s2)
.88                    ## 10. W

### Step 2: Quality Filtering

In [34]:
%%bash
time pyrad -p /data/analysis/params.txt -s 2

	/data/analysis/edits/1HL_17A.assembled already in edits/
	/data/analysis/edits/1HL_11A.assembled already in edits/
	/data/analysis/edits/1HL_10A.assembled already in edits/
	/data/analysis/edits/1HL_13A.assembled already in edits/
	/data/analysis/edits/1HL_12A.assembled already in edits/
	/data/analysis/edits/1HL_16A.assembled already in edits/
	/data/analysis/edits/1HL_15A.assembled already in edits/
	/data/analysis/edits/1HL_14A.assembled already in edits/




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------

	sorted .fastq from /data/analysis/fastq/*.assembled.fastq being used
	step 2: editing raw reads 
	........................................................................................
real	70m16.887s
user	453m14.412s
sys	1m16.616s


#### Let's take a look at the results

In [35]:
%%bash
cat /data/analysis/stats/s2.rawedit.txt

sample 	Nreads	passed	passed.w.trim	passed.total
1HL_10A.assembled	2890999	2127357	214294	2341651
1HL_11A.assembled	3088425	2242089	246561	2488650
1HL_12A.assembled	2323231	1728934	163227	1892161
1HL_13A.assembled	2701305	2075000	175929	2250929
1HL_14A.assembled	1865526	1407146	129089	1536235
1HL_15A.assembled	1837684	1342114	139898	1482012
1HL_16A.assembled	2100838	1568118	147901	1716019
1HL_17A.assembled	3106139	2259592	237499	2497091

    Nreads = total number of reads for a sample
    passed = retained reads that passed quality filtering at full length
    passed.w.trim= retained reads that were trimmed due to detection of adapters
    passed.total  = total kept reads of sufficient length
    note: you can set the option in params file to include trimmed reads of xx length. 
sample 	Nreads	passed	passed.w.trim	passed.total
1HL_19A.assembled	3174884	2334130	238750	2572880
1HL_1A.assembled	2098741	1574224	145441	1719665
1HL_20A.assembled	2516582	1852015	188703	2040718
1HL_21A.assembl

### Step 3: Clustering

In [36]:
%%bash
time pyrad -p /data/analysis/params.txt -s 3

	skipping 1HL_17A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_11A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_10A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_13A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_12A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_16A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_14A.assembled.clustS.gz already exists in /data/analysis/clust.88
	skipping 1HL_15A.assembled.clustS.gz already exists in /data/analysis/clust.88




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	de-replicating files for clustering...

	step 3: within-sample clustering of 96 samples at 
	        '.88' similarity. Running 8 parallel jobs
	 	with up to 16 threads per job. If needed, 
		adjust to avoid CPU and MEM limits

	sample 1NF_29A finished, 97131 loci
	sample 1NF_31A finished, 109056 loci
	sample 1NF_4A finished, 110918 loci
	sample 1NF_32A finished, 115584 loci
	sample 1SN_14A finished, 110527 loci
	sample 1SN_23A finished, 118344 loci
	sample 1SN_30A finished, 121564 loci
	sample 1NF_13A finished, 127575 loci
	sample 1HL_19A finished, 111346 loci
	sample 1HL_23A finished, 94229 loci
	sample 1NF_2A finished, 102160 loci
	sample 1SN_8A finished, 102124 loci
	sample 1NF_23A finished, 122732 loci
	sample 1HL_28A finished, 104605 loci
	sample 1SN_21A finished, 131359 loci
	sample

#### Let's take a look at the results

In [37]:
%%bash
cat /data/analysis/stats/s3.clusters.txt


taxa	total	dpt.me	dpt.sd	d>5.tot	d>5.me	d>5.sd	badpairs

    ## total = total number of clusters, including singletons
    ## dpt.me = mean depth of clusters
    ## dpt.sd = standard deviation of cluster depth
    ## >N.tot = number of clusters with depth greater than N
    ## >N.me = mean depth of clusters with depth greater than N
    ## >N.sd = standard deviation of cluster depth for clusters with depth greater than N
    ## badpairs = mismatched 1st & 2nd reads (only for paired ddRAD data)

HISTOGRAMS

    

taxa	total	dpt.me	dpt.sd	d>5.tot	d>5.me	d>5.sd	badpairs

    ## total = total number of clusters, including singletons
    ## dpt.me = mean depth of clusters
    ## dpt.sd = standard deviation of cluster depth
    ## >N.tot = number of clusters with depth greater than N
    ## >N.me = mean depth of clusters with depth greater than N
    ## >N.sd = standard deviation of cluster depth for clusters with depth greater than N
    ## badpairs = mismatched 1st & 2nd reads (only for p

### Step 4: Parameter estimation

In [38]:
%%bash
time pyrad -p /data/analysis/params.txt -s 4



     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 4: estimating error rate and heterozygosity
	................................................................................................
real	1033m58.941s
user	7810m54.513s
sys	0m26.031s


In [1]:
%%bash
cat /data/analysis/stats/Pi_E_estimate.txt

taxa	H	E
1HL_13A.assembled	0.01234178	0.00144187	
1HL_10A.assembled	0.01399484	0.00139753	
1HL_15A.assembled	0.01242723	0.0011879	
1HL_12A.assembled	0.01240698	0.00157456	
1HL_16A.assembled	0.01381188	0.00054237	
1HL_17A.assembled	0.01332716	0.00127763	
1HL_14A.assembled	0.01266762	0.00203182	
1HL_11A.assembled	0.01216978	0.00165352	
1HL_13A.assembled	0.01234178	0.00144187	
1HL_10A.assembled	0.01399484	0.00139753	
1HL_15A.assembled	0.01242723	0.0011879	
1HL_12A.assembled	0.01240698	0.00157456	
1HL_16A.assembled	0.01381188	0.00054237	
1HL_17A.assembled	0.01332716	0.00127763	
1HL_14A.assembled	0.01266762	0.00203182	
1NF_1A.assembled	0.01302939	0.00167192	
1NF_19A.assembled	0.01302296	0.00156732	
1SN_24A.assembled	0.01280068	0.00172812	
1NF_18A.assembled	0.01298092	0.00169576	
1SN_29A.assembled	0.01288545	0.0014629	
1HL_6A.assembled	0.01312247	0.00152568	
1NF_25A.assembled	0.01280702	0.00162634	
1NF_12A.assembled	0.01233271	0.00160106	
1HL_5A.assembled	0.01210494	0.00151496	
1NF_10A.assem

### Step 5: Consenus base calling

In [2]:
%%bash
time pyrad -p /data/analysis/params.txt -s 5



     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 5: creating consensus seqs for 96 samples, using H=0.01247 E=0.00154
	................................................................................................
real	283m28.123s
user	2225m48.454s
sys	0m21.896s


In [3]:
%%bash
cat /data/analysis/stats/s5.consens.txt

taxon          	nloci	f1loci	f2loci	nsites	npoly	poly
1SN_9A.assembled	126666	39351	36531	3608975	27926	0.0077379
1NF_21A.assembled	161246	39443	36671	3636146	28281	0.0077777
1HL_11A.assembled	125407	39736	36789	3659499	28748	0.0078557
1SN_21A.assembled	131359	40405	37391	3710615	29515	0.0079542
1NF_23A.assembled	122732	40195	37205	3685221	27872	0.0075632
1SN_23A.assembled	118344	43478	40280	4003823	30280	0.0075628
1SN_30A.assembled	121564	43820	40498	4028393	30831	0.0076534
1NF_13A.assembled	127575	43904	40669	4061429	31655	0.0077941
1SN_19A.assembled	120143	37039	34302	3391779	26288	0.0077505
1SN_5A.assembled	121723	38622	35831	3535911	27199	0.0076922
1SN_12A.assembled	116140	41408	38556	3826417	29483	0.0077051
1HL_28A.assembled	104605	38307	35400	3494480	29496	0.0084407
1NF_32A.assembled	115584	42091	39211	3910432	29914	0.0076498
1NF_4A.assembled	110918	42309	39302	3937693	30501	0.0077459
1HL_19A.assembled	111346	38543	35529	3516253	28770	0.008182
1SN_14A.assembled	110527	39960	3701

### Cluster across samples

In [4]:
%%bash
time pyrad -p /data/analysis/params.txt -s 6

vsearch v1.11.1_linux_x86_64, 14.7GB RAM, 8 cores
https://github.com/torognes/vsearch


	finished clustering




     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	step 6: clustering across 96 samples at '.88' similarity 

Reading file /data/analysis/clust.88/cat.haplos_ 100%
291106371 nt in 2830448 seqs, min 32, max 234, avg 103
Counting unique k-mers 100%
Clustering 100%
Sorting clusters 100%
Writing clusters 100%
Clusters: 125434 Size min 1, max 2641, avg 22.6
Singletons: 25519, 0.9% of seqs, 20.3% of clusters

real	38m13.548s
user	60m4.549s
sys	0m14.356s


### Assemble final data set

In [5]:
%%bash
time pyrad -p /data/analysis/params.txt -s 7

	ingroup 1HL_10A.assembled,1HL_11A.assembled,1HL_12A.assembled,1HL_13A.assembled,1HL_14A.assembled,1HL_15A.assembled,1HL_16A.assembled,1HL_17A.assembled,1HL_19A.assembled,1HL_1A.assembled,1HL_20A.assembled,1HL_21A.assembled,1HL_22A.assembled,1HL_23A.assembled,1HL_24A.assembled,1HL_25A.assembled,1HL_26A.assembled,1HL_27A.assembled,1HL_28A.assembled,1HL_29A.assembled,1HL_2A.assembled,1HL_31A.assembled,1HL_33A.assembled,1HL_34A.assembled,1HL_35A.assembled,1HL_3A.assembled,1HL_4A.assembled,1HL_5A.assembled,1HL_6A.assembled,1HL_7A.assembled,1HL_8A.assembled,1HL_9A.assembled,1NF_10A.assembled,1NF_11A.assembled,1NF_12A.assembled,1NF_13A.assembled,1NF_14A.assembled,1NF_15A.assembled,1NF_16A.assembled,1NF_17A.assembled,1NF_18A.assembled,1NF_19A.assembled,1NF_1A.assembled,1NF_20A.assembled,1NF_21A.assembled,1NF_22A.assembled,1NF_23A.assembled,1NF_24A.assembled,1NF_25A.assembled,1NF_26A.assembled,1NF_27A.assembled,1NF_28A.assembled,1NF_29A.assembled,1NF_2A.assembled,1NF_30A.assembled,1NF_31A.asse



     ------------------------------------------------------------
      pyRAD : RADseq for phylogenetics & introgression analyses
     ------------------------------------------------------------


	Cluster input file: using 
	/data/analysis/clust.88/cat.clust_.gz

........Traceback (most recent call last):
  File "/usr/local/bin/pyrad", line 9, in <module>
    load_entry_point('pyrad==3.0.66', 'console_scripts', 'pyrad')()
  File "build/bdist.linux-x86_64/egg/pyrad/pyRAD.py", line 527, in main
  File "build/bdist.linux-x86_64/egg/pyrad/alignable.py", line 1013, in main
  File "build/bdist.linux-x86_64/egg/pyrad/loci2mig.py", line 75, in make
IndexError: list index out of range

real	62m40.523s
user	199m9.651s
sys	18m54.962s


#### Well, that error is a bit disconcerting... Let's look at the results and output files.

### Examine results

In [6]:
%%bash
cat /data/analysis/stats/oly_gbs_pyrad.stats



71174       ## loci with > minsp containing data
46377       ## loci with > minsp containing data & paralogs removed
46377       ## loci with > minsp containing data & paralogs removed & final filtering

## number of loci recovered in final data set for each taxon.
taxon	nloci
1HL_10A.assembled	175
1HL_11A.assembled	13726
1HL_12A.assembled	250
1HL_13A.assembled	89
1HL_14A.assembled	1124
1HL_15A.assembled	242
1HL_16A.assembled	253
1HL_17A.assembled	614
1HL_19A.assembled	13237
1HL_1A.assembled 	9589
1HL_20A.assembled	10722
1HL_21A.assembled	10058
1HL_22A.assembled	8946
1HL_23A.assembled	13897
1HL_24A.assembled	10120
1HL_25A.assembled	11994
1HL_26A.assembled	12644
1HL_27A.assembled	13331
1HL_28A.assembled	13299
1HL_29A.assembled	11531
1HL_2A.assembled 	11332
1HL_31A.assembled	10212
1HL_33A.assembled	10466
1HL_34A.assembled	11384
1HL_35A.assembled	10929
1HL_3A.assembled 	9789
1HL_4A.assembled 	10970
1HL_5A.assembled 	8488
1HL_6A.assembled 	8568
1HL_7A.assembled 	11704
1HL_8A.assembled 	9

#### I'm not sure what all of that means...

### Output files

In [7]:
%%bash
ls /data/analysis/outfiles/

oly_gbs_pyrad.alleles
oly_gbs_pyrad.excluded_loci
oly_gbs_pyrad.gphocs
oly_gbs_pyrad.loci
oly_gbs_pyrad.migrate
oly_gbs_pyrad.nex
oly_gbs_pyrad.phy
oly_gbs_pyrad.phy.partitions
oly_gbs_pyrad.snps
oly_gbs_pyrad.snps.geno
oly_gbs_pyrad.str
oly_gbs_pyrad.treemix.gz
oly_gbs_pyrad.unlinked_snps
oly_gbs_pyrad.usnps.geno
oly_gbs_pyrad.vcf
